Hypervisor list

baneofblabs opened the pull request:
Add a panel to display a grid of hypervisors, each one showing which instances are running on it. Originally this was written to stand alone, and only later made into a dashboard panel, which probably shows. I hadn't made a dashboard panel before, and found documentation limited, so I expect I'll have done some things suboptimally...

baneofblabs committed:
Add panel for visualising hypervisor utilisation.

baneofblabs committed:
Fix bug caused by hypervisors with no instances.

Everybody loves bounds checking.

baneofblabs committed:
Fix bug using wrong attribute for host name.

baneofblabs committed:
Move javascript code inline.

This is to work around a javascript error that was appearing when the
code was included as before. Not sure how to fix it properly...

baneofblabs committed:
Merge branch 'nci/kilo' into hvlist

baneofblabs committed:
Be more defensive.

Check array bounds, report errors, don't trust openstack.

chrisacbr commented on openstack_dashboard/local/dashboards/admin_nci/hvlist/views.py:
```diff
@@ -0,0 +1,160 @@
+                host = getattr(i, 'OS-EXT-SRV-ATTR:host')
+            except AttributeError:
+                # this will make a pretty box for the error message, although it assumes it's an HTTP error so writes "HTTP None"
+                raise ClientException(None, 'could not get OS-EXT-SRV-ATTR:host attribute of instance')
```
Do you really want to bail out completely here?  The other instances might be OK...  Perhaps "messages.error()" and continue next iteration.  Similarly for flavours/projects below.

chrisacbr commented on openstack_dashboard/local/dashboards/admin_nci/hvlist/views.py:
```diff
@@ -0,0 +1,160 @@
+                # (wrap this rather trivially in a try block to make the error less cryptic)
+                try:
+                    flavs[i.flavor['id']] = api.nova.flavor_get(request, i.flavor['id'])
+                    messages.warning(request, 'Extra lookup for flavor "'+str(i.flavor['id'])+'"')
```
I think maybe a warning pop-up for this case might be a bit annoying.  You've still got the info just via another route so simply continue silently (or use "LOG.debug")?  Same for project below.

chrisacbr commented on openstack_dashboard/local/dashboards/admin_nci/hvlist/views.py:
```diff
@@ -0,0 +1,160 @@
+                s.flav.description = '{cpu}/{mem}/{disk}'.format(
+                    cpu  = s.flav.vcpus,
+                    mem  = int(s.flav.ram*1024**2*mem_scale+0.5),
+                    disk = int(s.flav.disk*1024**3*disk_scale+0.5),
```
You also need to add  "OS-FLV-EXT-DATA:ephemeral" to the disk usage.

chrisacbr commented on openstack_dashboard/local/dashboards/admin_nci/hvlist/views.py:
```diff
@@ -0,0 +1,160 @@
+        context['hypervisors'] = hypervisors
+        context['used_count'] = sum(1 for h in hypervisors if h.servers)
+        context['server_count'] = sum(len(h.servers) for h in hypervisors)
+        context['cluster_name'] = 'Devstack'
```
Not used anywhere?

baneofblabs committed:
Address feedback from Chris.

Don't bail completely if an instance is missing OS-EXT-SRV-ATTR:host,
but instead just ignore it, show an error, and try to show the remaining
instances.

When details for a flavor or project need to be fetched a second time,
log a warning instead of showing a pop-up warning message, to be less
annoying...

Include ephemeral disk usage in calculations.

Remove some redundant code and also clean up a little bit.

baneofblabs committed:
Improve display of flavors.

chrisacbr commented:
There's still a couple of `raise NotFound()` in there.  Sure you don't want to use `messages.error()` instead there too?

baneofblabs committed:
Don't give up when seeing unknown flavor/tenant.

Keep calm and display an error message, instead of throwing an exception
and dying. The inaccuracy of the output will be kind of proportional to
the number of error messages displayed...

baneofblabs committed:
Improve display.

chrisacbr commented:
I've put `tc096` and `tc097` into two host aggregates like we have in production so you can see a case where a host is and isn't overcommitting vCPUs.  At the moment this is causing the CPU bar to wrap around to the next row which is a purely cosmetic issue.

Not entirely sure what the intention with this panel is though.  If you want to show usage with respect to actual physical resources then the current approach will certainly achieve that.  Although that will mean hosts will often be shown as red in aggregates where we are overcommitting even though they aren't near the vCPU limit.  OTOH, if you want to use the overcommit (or perhaps even undercommit) limits then you'll need to:
- Retrieve the list of host aggregates with associated metadata and assigned hosts.
- For each hypervisor:
  - Determine if the host is a member of an aggregate.
  - For each resource, check if the aggregate metadata defines `cpu_allocation_ratio`, `ram_allocation_ratio` or `disk_allocation_ratio` as appropriate and use the value as a multiplier of the physical limit.
  - If there is no aggregate specific ratio for a given resource, fall back to the global ratios.  These are defined in `/etc/nova/nova.conf`.  Reading that file directly isn't ideal but the information doesn't appear to be exposed via the Nova API.
- Would be a good idea to display the hosts grouped by aggregate and sort hosts by name within each group.

baneofblabs committed:
Group hosts by host aggregates.

baneofblabs committed:
Fix display of overcommitted resources.

Stop the indicator bars from overflowing, and display a red border when
resources are overcommitted.

This doesn't yet take into account overcommit ratios; it's just counting
physical resources.

baneofblabs committed:
Refactor!

This code was really old and ugly... now some bits aren't as old. :)

baneofblabs committed:
Account for overcommit limits.

baneofblabs committed:
Don't die if nova.conf can't be read.

baneofblabs committed:
Fix "total hypervisors" bug.

baneofblabs committed:
Fix 'fix "total hypervisors" bug' bug.

baneofblabs committed:
Use django settings for overcommit ratios.
